segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ThrottlerCalculator.java

```diff
@@ -366,7 +437,11 @@ public String toString() {
         /**
          * Throttling is required due to excessive size of the Commit (Memory) Backlog Queue.
          */
-        CommitBacklog
+        CommitBacklog,
+        /**
+         * Throttlign is required due to excessive size of DurableDataLog's in-flight queue.
```

Comment:
Typo

Reply:
Thanks @eolivelli, yes we tried configuring the `setThrottleValue()` parameter to low numbers (e.g., 100, 10), but it is not preventing the system to be unstable in cases of "unthrottled" clients and low IO performance Tier 1 storage (i.e., where Bookkeeper writes to). I'm validating this PR, and it looks like it is making the system stable even in non-favorable scenarios as the one mentioned. But note that this does not prevent Pravega from benefiting from the throttling capabilities of newer versions of Bookkeeper (if I'm not wrong, in `4.8` there was a server-side throttling mechanism implemented).

Reply:
@eolivelli Thank you for your suggestion. Like @RaulGracia said, that was one of the first things we tried. However I noticed that it does little to prevent backpressure on the application itself and simply guards how many entries it has "in-flight" to the bookies. At least based on the latest code in the BK repository, this uses Guava's `RateLimiter` which blocks synchronously until permits are available. In Pravega's situation that would quickly exhaust all of our threads and lead to a lot of problems, so we need an async solution.

Pravega's SegmentStore already has an async throttler that can be configured with various inputs (triggers and "cancelers") which implement a gradual throttling scheme based on how busy the system is (two thresholds: min threshold is when we throttle a little bit, and max threshold is when we apply a full stop). The advantage of this async throttler is that it doesn't block any threads in doing so, and those threads can be used by other components of the system to clear out backlog or perform other necessary functions.